### PR TITLE
Implement HTTP endpoint factory

### PR DESCRIPTION
**Pull request by hacksparrow**
Implements #1331.

Main changes:

1. Added a new package `@loopback/http-server`
2. Made changes to `@loopback/rest` to use `@loopback/http-server`. 

## Checklist

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

### CODE REVIEW

File: packages/http-server/src/http-server.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  /**
+   * Starts the HTTP / HTTPS server
+   */
+  public start(): Promise<void> {
+    this.httpServer.listen(this.httpPort, this.httpHost);
+    return new Promise<void>(async (resolve, reject) => {
+      try {
+        await pEvent(this.httpServer, 'listening');
+        const address = this.httpServer.address() as AddressInfo;
+        this.restServer.bind('rest.port').to(address.port);
```
Comment (hacksparrow) [inline packages/http-server/src/http-server.ts]: @bajtos  I wanted to use `RestBindings.PORT`, but `RestBindings` itself is undefined when `{RestBindings}` is imported into `http-sever.ts` - `http-sever.ts` is executed before the `RestBindings` object is created.

If there is a solution, do let me know.

File: packages/http-server/src/http-server.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+ * HTTP / HTTPS server used by LoopBack's RestServer
+ *
+ * @export
+ * @class HttpServer
+ */
+export class HttpServer {
+  private restServer: RestServer;
+  private httpPort: number;
+  private httpHost: string | undefined;
+  private httpServer: Server;
```
Comment (bajtos) [inline packages/http-server/src/http-server.ts]: Is it necessary to prefix all variables with `http`, considering that they belong to a class that already has `http` in the name?

File: packages/http-server/src/http-server.ts (hunk 1 above)
Comment (bajtos) [reply packages/http-server/src/http-server.ts]: IMO, `HttpServer` should not deal with dependency injection at all. Instead, it should provide properties like `port`, `hostname` and perhaps `url` that can be used by consumers to update any bindings as necessary.

```ts
class HttpServer {
  get port(): number {
    return this.address().port;
  }

  address: AddressInfo {
    return this.address() as AddressInfo;
  }
}
```

Later in RestServer:

```js
async start() {  
  // create http server and start it, then:
  this.bind(RestBindings.PORT).to(server.port);
}
```

File: packages/http-server/src/http-server.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    this.httpHost = httpOptions.host;
+    this.httpServer = createServer(httpRequestListener);
+  }
+
+  /**
+   * Starts the HTTP / HTTPS server
+   */
+  public start(): Promise<void> {
+    this.httpServer.listen(this.httpPort, this.httpHost);
+    return new Promise<void>(async (resolve, reject) => {
```
Comment (bajtos) [inline packages/http-server/src/http-server.ts]: What is the purpose of this `new Promise` wrapper? Just make the `start` method async.

```ts
public async start(): Promise<void> {
  this.httpServer.listen(...);
  await pEvent(this.httpServer, 'listening');
}
```

File: packages/http-server/src/http-server.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+      }
+    });
+  }
+
+  /**
+   * Stops the HTTP / HTTPS server
+   */
+  public stop(): Promise<void> {
+    this.httpServer.close();
+    return new Promise<void>(async (resolve, reject) => {
```
Comment (bajtos) [inline packages/http-server/src/http-server.ts]: Again, no need for creating a new promise, just make `stop()` an async function (method).

File: packages/http-server/src/http-server.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+      try {
+        await pEvent(this.httpServer, 'close');
+        resolve();
+      } catch (e) {
+        reject();
```
Comment (bajtos) [inline packages/http-server/src/http-server.ts]: This silently discards the error, which is very evil! The correct solution is `reject(e)`. Does not really matter here because this whole try/catch block is not needed, but it's important to take more care in the future.

File: packages/http-server/test/integration/http-server.integration.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: @loopback/http-server
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+import {RestServer, RestComponent} from '@loopback/rest';
+import {Application, ApplicationConfig} from '@loopback/core';
```
Comment (bajtos) [inline packages/http-server/test/integration/http-server.integration.ts]: http-server should not depend on `@loopback/core` and definitely not depend on `@loopback/rest`.

The integration tests should test `HttpServer` by calling its API, the tests are integration because under the hood, a real HTTP server will be created (integration with the system APIs is tested).

File: packages/http-server/test/integration/http-server.integration.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+import * as assert from 'assert';
+
+describe('HttpServer', () => {
+  it('starts server', async () => {
+    const server = await givenAServer();
+    await server.start();
+    assert(server.listening, 'Server not started');
```
Comment (bajtos) [inline packages/http-server/test/integration/http-server.integration.ts]: This does not really verify that a server is started. You need to make an HTTP request and verify that it was handled as expected. Please use supertest as the client, perhaps by using `@loopback/testlab` - see https://github.com/strongloop/loopback-next/blob/master/packages/testlab/src/client.ts#L11-L13.

We also don't use `assert` from Node.js core because it has poor ergonomic and unhelpful error messages. Please use `expect` provided by testlab, see https://github.com/strongloop/loopback-next/tree/master/packages/testlab#basic-use and http://shouldjs.github.io

File: packages/http-server/test/integration/http-server.integration.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    await server.stop();
+  });
+
+  it('stops server', async () => {
+    const server = await givenAServer();
+    await server.start();
+    await server.stop();
```
Comment (bajtos) [inline packages/http-server/test/integration/http-server.integration.ts]: Similarly here - I think you need to verify that the server is no longer listening on the given port, e.g. a request fails.

File: packages/http-server/test/unit/http-server.unit.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: @loopback/http-server
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+import {HttpServer} from '../..';
+import {RestServer} from '@loopback/rest';
+import {Application} from '@loopback/core';
+import * as assert from 'assert';
+
+describe('HttpServer', () => {
```
Comment (bajtos) [inline packages/http-server/test/unit/http-server.unit.ts]: AFAICT, these tests are just duplicating integration tests - what's the point?

IMO, unit tests for an http server class makes little sense. You would have to mock core HTTP module, I think the complexity of that is bigger than the benefits of testing in isolation.

I am proposing to remove unit tests completely and move any relevant bits to integration tests.

File: packages/rest/src/rest.server.ts [hunk 10]
```diff
@@ -167,6 +166,15 @@ export class RestServer extends Context implements Server, HttpServerLike {
     this._setupRequestHandler(options);
 
     this.bind(RestBindings.HANDLER).toDynamicValue(() => this.httpHandler);
+
+    this._httpServer = new HttpServer(
```
Comment (bajtos) [inline packages/rest/src/rest.server.ts]: We need to defer creation of HttpServer instance until `start` is called, so that we can pick any changes in HOST/PORT bindings.

An outline of a test case to write:

```ts
const server = new RestServer({port: 80});
server.bind(RestBindings.PORT).to(0);
await server.start();
// get the actual port number and verify it's not 80
```

File: packages/rest/src/rest.server.ts [hunk 11]
```diff
@@ -572,24 +580,7 @@ export class RestServer extends Context implements Server, HttpServerLike {
     // Setup the HTTP handler so that we can verify the configuration
     // of API spec, controllers and routes at startup time.
     this._setupHandlerIfNeeded();
-
-    const httpPort = await this.get<number>(RestBindings.PORT);
-    const httpHost = await this.get<string | undefined>(RestBindings.HOST);
-    this._httpServer = createServer(this.requestHandler);
```
Comment (bajtos) [inline packages/rest/src/rest.server.ts]: Continuing my previous comment - this is the place where to instantiate `HttpServer` as a replacement for `http.createServer`. We need to preserve the logic obtaining the latest values for port & host.

File: packages/http-server/README.md [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+`@loopback/http-server` should be instantiated with an instance of `RestServer`, HTTP / HTTPS options object, and a request handler function.
+
+```js
+import {RestServer} from '@loopback/rest';
+import {Application} from '@loopback/core';
+
+const app = new Application();
+const restServer = new RestServer(app);
+const httpServer = new HttpServer(restServer, {port: 3000, host: ''}, (req, res) => {});
```
Comment (bajtos) [inline packages/http-server/README.md]: This example will need to be updated once we remove the dependency from HttpServer to RestServer, see my comments below.

I think it also makes more sense to reorder constructor arguments to accept request handler as the first arg and options as the second.

I would like something along the following lines:

```ts
import {HttpServer} from '@loopback/http-server';

const httpServer = new HttpServer(
  (req, res) => res.end('Hello world'),
  {port: 3000}
);
```

File: packages/http-server/test/unit/http-server.unit.ts (hunk 9 above)
Comment (hacksparrow) [reply packages/http-server/test/unit/http-server.unit.ts]: > I am proposing to remove unit tests completely and move any relevant bits to integration tests.

Yes, totally agree. I didn't feel OK as I was writing those tests.

File: packages/http-server/package.json [hunk 13]
```diff
@@ -0,0 +1,50 @@
+{
+  "name": "@loopback/http-server",
+  "version": "0.0.0",
```
Comment (jannyHou) [inline packages/http-server/package.json]: I think  the first version would be "0.0.1"?

**Comment by raymondfeng** [comment]
I don't know why we completely threw away the abstraction at https://github.com/strongloop/loopback-next/blob/2bddc47ccda56a50c52a8a92facd87e3a3f4944d/packages/http-server/src/common.ts. IMO, the PoC only requires some clean up to align with latest code in master. 

I agree with @bajtos's comments that `http-server` should be only responsible for creating/starting/stopping http endpoints and exposing server address. No dependency on core/rest modules should exist.

**Comment by hacksparrow** [comment]
@raymondfeng, @bajtos suggested to keep this PR as simple as possible and build from there.

**Comment by raymondfeng** [comment]
> @raymondfeng, @bajtos suggested to keep this PR as simple as possible and build from there.

I thought the idea was to decompose my original PR into a few steps and refine each of them as we go.  If we can combine `http-server` and `http-server-express` from the PoC, realign with master, trim/improve the code and test coverage, it should have been more effective. The key is to provide a good divide of responsibility so that other modules don't have to deal with lower-level/common http processing, which includes:

- Create http/https/... endpoints based on the config
- Manage the lifecycle of http endpoints (start/stop)
- Expose bound addresses
- Marshaling/unmarshaling
- Middleware

**Comment by bajtos** [comment]
@raymondfeng 

> > @raymondfeng, @bajtos suggested to keep this PR as simple as possible and build from there.
>
> I thought the idea was to decompose my original PR into a few steps and refine each of them as we go. 

Yes, that's my idea too!

My general opinion is that we should treat spikes/PoCs as throw-away code and build the production implementation from scratch using test-driven/test-first development. Otherwise it's too easy to overlook that some parts of spike code that leaked into production are not in production quality and/or don't have enough test coverage. 

In this particular case, the PoC was created with the assumption that we are going to support multiple server frameworks (Koa, Express), which introduced additional complexity to the API. For example, consider `HttpFactory<REQ, RES, APP>` - do we still need `REQ` and `RES` generic/template parameters?

So far, we were following the plan outlined in https://github.com/strongloop/loopback-next/pull/1082#issuecomment-384580197:

> 1. A pair of (req, res) arguments was changed to a single httpContext object.
> 2. Rework of REST internals to support pluggable HTTP transports (Express, Koa, etc.)
> 3. Introduction of Endpoint classes that encapsulate creation of HTTP/HTTPs servers, etc.

Since there was no follow-up comment from you, we assumed you are ok with my proposal 🤷‍♂️ 

> The key is to provide a good divide of responsibility so that other modules don't have to deal with lower-level/common http processing, which includes:
> - Create http/https/... endpoints based on the config
> -Manage the lifecycle of http endpoints (start/stop)
> - Expose bound addresses
> - Marshaling/unmarshaling
> - Middleware

This is a good list that allows me to better understand your vision. I don't remember seeing such list before, it would be great to have it at a prominent place in your PoC PR or some other place where its easy to find. Having said that, the fault may be at my side too, these changes are in progress for more than a month.

Back to the topic: The first three items make perfect sense and I think that's exactly the scope of this pull request. I have concerns about putting marshaling/unmarshaling and middleware capabilities into our shared low-level HTTP stack. I'd like us to work on these two changes in a different pull request(s), so that the initial set of features can be landed faster, and then we can keep discussions about the remaining two items focused and in a single place.

Thoughts?

File: packages/rest/src/rest.server.ts [hunk 14]
```diff
@@ -600,16 +591,7 @@ export class RestServer extends Context implements Server, HttpServerLike {
...
-    return new Promise<void>((resolve, reject) => {
-      server.close((err: Error) => {
-        if (err) {
-          reject(err);
-        } else {
-          resolve();
-        }
-      });
-    });
+    return this._httpServer.stop();
```
Comment (bajtos) [inline packages/rest/src/rest.server.ts]: Now that you are directly returning the promise provided by `stop`, this function no longer needs to be async. 

```diff
-   async stop() {
+  stop() {
        // Kill the server instance.
```

File: packages/http-server/src/http-server.ts [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+   */
+  public async stop() {
+    this.httpServer.close();
+    await pEvent(this.httpServer, 'close');
+  }
+
+  /**
+   * Protocol of the HTTP / HTTPS server
+   */
+  public get protocol(): 'http' | 'https' {
```
Comment (bajtos) [inline packages/http-server/src/http-server.ts]: Please define a shared type for `'http' | 'https'` and use it all places where you are duplicating this union now.

E.g.

```ts
export type HttpProtocol = 'http' | 'https';
```

File: packages/http-server/src/http-server.ts [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+   * Starts the HTTP / HTTPS server
+   */
+  public async start() {
+    this.httpServer = createServer(this.httpRequestListener);
+    this.httpServer.listen(this._port, this._host);
+    await pEvent(this.httpServer, 'listening');
+    this._address = this.httpServer.address() as AddressInfo;
+    this._host = this._host || this._address.address;
+    this._port = this._address.port;
+    this._protocol = this._protocol || 'http';
```
Comment (bajtos) [inline packages/http-server/src/http-server.ts]: Please move this initialization to the constructor.

File: packages/rest/src/rest.server.ts (hunk 14 above)
Comment (hacksparrow) [reply packages/rest/src/rest.server.ts]: Will await `this._httpServer.stop()`, instead.

File: packages/http-server/src/http-server.ts [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+    if (this._address.family === 'IPv6') {
+      this._host = `[${this._host}]`;
```
[the remark this answers is not in the document]
Comment (hacksparrow) [reply packages/http-server/src/http-server.ts]: Ok, let's address the format when and where encountered. Will remove the formatting code.

File: packages/http-server/test/integration/http-server.integration.ts [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+    const app = new Application({
+      rest: {
+        port: 0,
+      },
+    });
+    app.component(RestComponent);
+    const server = await app.getServer(RestServer);
+    server.bind(RestBindings.PORT).to(0);
+    await server.start();
+    // @bajtos `_httpServer` of `server` is protected, how do we check the port and other details here?
```
[the remark this answers is not in the document]
Comment (hacksparrow) [reply packages/http-server/test/integration/http-server.integration.ts]: @bajtos that's a `RestServer ` instance. `HttpServer` does not do any sort of binding now, it just starts a server and exposes properties about that server.

[the remark this answers is not in the document]
Comment (bajtos) [reply packages/http-server/test/integration/http-server.integration.ts]: I see. Since HttpServer does not deal with any bindings, this test does not belong to `packages/rest`. `packages/rest` is the place where to test this functionality - I would be surprised if there wasn't any existing test for that.

File: packages/rest/src/rest.server.ts (hunk 11 above)
Comment (bajtos) [reply packages/rest/src/rest.server.ts]: Cross-posting https://github.com/strongloop/loopback-next/pull/1369#discussion_r191401296

An outline of a test case to write:

```ts
const server = new RestServer({port: 80});
server.bind(RestBindings.PORT).to(0);
await server.start();
// get the actual port number and verify it's not 80
```

File: packages/http-server/src/http-server.ts [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+  req: ServerRequest,
+  res: ServerResponse,
+) => void;
+
+/**
+ * Object for specifyig the HTTP / HTTPS server options
+ */
+export type HttpServerOptions = {
+  port: number;
+  host?: string | undefined;
```
Comment (bajtos) [inline packages/http-server/src/http-server.ts]: `host?` already allows `undefined`, no need to repeat that again in the type definition.

```ts
host?: string;
```

File: packages/rest/src/rest.server.ts [hunk 20]
```diff
@@ -167,6 +167,12 @@ export class RestServer extends Context implements Server, HttpServerLike {
     this._setupRequestHandler(options);
 
     this.bind(RestBindings.HANDLER).toDynamicValue(() => this.httpHandler);
+
+    this._httpServer = new HttpServer(this.requestHandler, {
+      port: options.port,
+      host: options.host,
+      protocol: options.protocol || 'http',
+    });
```
Comment (bajtos) [inline packages/rest/src/rest.server.ts]: Please move the initialization of `_httpServer` to `_setupRequestHandler` method below.

File: packages/http-server/test/integration/http-server.integration.ts [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+import {HttpServer} from '../../';
+import {supertest, expect} from '@loopback/testlab';
+import * as makeRequest from 'request-promise-native';
+
+describe('HttpServer (integration)', () => {
+  it('starts server', async () => {
+    const server = new HttpServer(
+      (req, res) => {
+        res.end();
+      },
```
Comment (bajtos) [inline packages/http-server/test/integration/http-server.integration.ts]: Please extract this request handler that's repeated in all tests into a shared helper function.

```ts
const server = new HttpServer(
  dummyRequestHandler,
  // ...
);
```

File: packages/http-server/test/integration/http-server.integration.ts [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+      {port: 0},
```
Comment (bajtos) [inline packages/http-server/test/integration/http-server.integration.ts]: Would it make sense to make the `port` setting optional and defaulting to `0`?

Intended usage:

```ts
const server = new  HttpServer(requestHandler);
```

It would also allow us to simplify `RestServerConfig` - instead of copying all interface members, we could leverage inheritance.

```ts
interface RestServerConfig extends HttpServerConfig {
  // ...
}
```

Thoughts?

**Comment by raymondfeng** [comment]
> In this particular case, the PoC was created with the assumption that we are going to support multiple server frameworks (Koa, Express), which introduced additional complexity to the API. For example, consider HttpFactory<REQ, RES, APP> - do we still need REQ and RES generic/template parameters?

Since we have settled on Express, the generic parameters are not needed any more. Such type of cleanup is what I had in mind for this task.

> Responsibility of the http module
- Create http/https/... endpoints based on the config (implemented by my PR, https://github.com/strongloop/loopback-next/blob/2bddc47ccda56a50c52a8a92facd87e3a3f4944d/packages/http-server-express/src/http-server-express.ts#L61)
- Manage the lifecycle of http endpoints (start/stop) (implemented by my PR, https://github.com/strongloop/loopback-next/blob/2bddc47ccda56a50c52a8a92facd87e3a3f4944d/packages/http-server/src/common.ts#L176)
- Expose bound addresses (implemented by my PR, )
- Marshaling/unmarshaling (TBD)
- Middleware (TBD)

File: packages/http-server/test/integration/http-server.integration.ts [hunk 23]
```diff
@@ -0,0 +1,98 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: @loopback/http-server
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+import {HttpServer} from '../../';
+import {Application, ApplicationConfig} from '@loopback/core';
+import {RestServer, RestBindings, RestComponent} from '@loopback/rest';
```
Comment (bajtos) [inline packages/http-server/test/integration/http-server.integration.ts]: Please remove any references to `@loopback/rest` to avoid a cyclic dependency.

File: packages/http-server/test/integration/http-server.integration.ts [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+      .to.have.property('address')
+      .which.is.an.Object();
+    await server.stop();
+  });
+
+  it('does not override binding', async () => {
+    const server = await givenARestServer({rest: {port: 80}});
+    await server.bind(RestBindings.PORT).to(0);
+    await server.start();
+    expect(server.getSync(RestBindings.PORT)).to.not.equal(80);
```
Comment (bajtos) [inline packages/http-server/test/integration/http-server.integration.ts]: Please move this test to `packages/rest`.

Comment (hacksparrow) [reply packages/http-server/test/integration/http-server.integration.ts]: Indeed, it belongs to `packages/rest`.

File: packages/http-server/src/http-server.ts [hunk 25]
```diff
@@ -0,0 +1,114 @@
...
+   * @param httpRequestListener
+   */
+  constructor(
+    httpRequestListener: HttpRequestListener,
+    httpServerOptions?: HttpServerOptions,
+  ) {
+    this._port = (httpServerOptions && httpServerOptions.port) || 0;
+    this._host = (httpServerOptions && httpServerOptions.host) || undefined;
+    this._protocol =
+      (httpServerOptions && httpServerOptions.protocol) || 'http';
```
Comment (bajtos) [inline packages/http-server/src/http-server.ts]: [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).

```ts
if (!httpServerOptions) httpServerOptions = {};
this._port = httpServerOptions.port;
// ...
```

I think we can also shorten the argument name to simply `options`.

File: packages/http-server/src/http-server.ts [hunk 26]
```diff
@@ -0,0 +1,114 @@
...
+    this.httpRequestListener = httpRequestListener;
+  }
+
+  /**
+   * Starts the HTTP / HTTPS server
+   */
+  public async start(port?: number, host?: string) {
```
Comment (bajtos) [inline packages/http-server/src/http-server.ts]: Let's keep `start` parameterless please - both host and port should be configured at the time when the server is created. Unless there is a good justification why we need these two parameters?

File: packages/http-server/src/http-server.ts (hunk 25 above)
Comment (hacksparrow) [reply packages/http-server/src/http-server.ts]: Indeed!

**Comment by hacksparrow** [comment]
@raymondfeng All checks passed. Any comments?

**Comment by raymondfeng** [comment]
please investigate https://travis-ci.org/strongloop/loopback-next/jobs/388859072

File: packages/http-server/src/http-server.ts [hunk 27]
```diff
@@ -0,0 +1,112 @@
...
+    this._address = this.httpServer.address() as AddressInfo;
+    this._host = this._host || this._address.address;
+    this._port = this._address.port;
+  }
+
+  /**
+   * Stops the HTTP / HTTPS server
+   */
+  public async stop() {
+    this.httpServer.close();
```
Comment (raymondfeng) [inline packages/http-server/src/http-server.ts]: We should check if `this.httpServer` is set.

File: packages/http-server/src/http-server.ts [hunk 28]
```diff
@@ -0,0 +1,112 @@
...
+    await pEvent(this.httpServer, 'close');
+  }
+
+  /**
+   * Protocol of the HTTP / HTTPS server
+   */
+  public get protocol(): HttpProtocol {
```
Comment (raymondfeng) [inline packages/http-server/src/http-server.ts]: Why bother to have `getter` for these properties? Do you want to enforce `readonly`?

**Comment by raymondfeng** [comment]
Please also fix https://travis-ci.org/strongloop/loopback-next/jobs/388859076

File: packages/http-server/src/http-server.ts (hunk 28 above)
Comment (raymondfeng) [reply packages/http-server/src/http-server.ts]: Have you tried `readonly` keyword from TypeScript?

File: packages/http-server/src/http-server.ts [hunk 29]
```diff
@@ -0,0 +1,114 @@
...
+    this._host = this._host || this._address.address;
+    this._port = this._address.port;
+  }
+
+  /**
+   * Stops the HTTP / HTTPS server
+   */
+  public async stop() {
+    if (this.httpServer) {
+      this.httpServer.close();
```
Comment (raymondfeng) [inline packages/http-server/src/http-server.ts]: Should we expose the state of the `HttpServer`, such as `started: boolean`?

File: packages/http-server/src/http-server.ts [hunk 30]
```diff
@@ -0,0 +1,114 @@
...
+    this._protocol = httpServerOptions.protocol || 'http';
+  }
+
+  /**
+   * Starts the HTTP / HTTPS server
+   */
+  public async start() {
+    this.httpServer = createServer(this.httpRequestListener);
+    this.httpServer.listen(this._port, this._host);
+    await pEvent(this.httpServer, 'listening');
```
Comment (raymondfeng) [inline packages/http-server/src/http-server.ts]: I wonder why we don't use a callback for `httpServer.listen()` and `close()`.

File: packages/http-server/src/http-server.ts (hunk 28 above)
Comment (hacksparrow) [reply packages/http-server/src/http-server.ts]: Thanks, will use that.

File: packages/http-server/src/http-server.ts (hunk 29 above)
Comment (hacksparrow) [reply packages/http-server/src/http-server.ts]: Good suggestion, will add. Initially I had `listening`, but removed it.

File: packages/http-server/src/http-server.ts (hunk 30 above)
Comment (hacksparrow) [reply packages/http-server/src/http-server.ts]: Good idea, will add.

File: packages/http-server/src/http-server.ts [hunk 31]
```diff
@@ -0,0 +1,114 @@
...
+   */
+  public get host(): string | undefined {
+    return this._host;
+  }
+
+  /**
+   * URL of the HTTP / HTTPS server
+   */
+  public get url(): string {
+    return `${this._protocol}://${this.host}:${this.port}`;
```
Comment (raymondfeng) [inline packages/http-server/src/http-server.ts]: Please make it more robust per https://github.com/strongloop/loopback-next/blob/2bddc47ccda56a50c52a8a92facd87e3a3f4944d/packages/http-server/src/common.ts#L176-L201

File: packages/http-server/src/http-server.ts (hunk 30 above)
Comment (bajtos) [reply packages/http-server/src/http-server.ts]: Please note that `pEvent` will listen for `error` events too. As a result, `server.start()` returns a rejected promise if the server cannot be started.

The callback for `httpServer.listen` is not error-first, it's just a regular event callback for `listening` event. See https://nodejs.org/dist/latest-v10.x/docs/api/net.html#net_server_listen:

> When the server starts listening, the 'listening' event will be emitted. The last parameter callback will be added as a listener for the 'listening' event.

@hacksparrow Please start by adding a test to verify that `start()` returns a rejected promise when the server cannot be started. Quoting from the same doc page:

> One of the most common errors raised when listening is `EADDRINUSE`. This happens when another server is already listening on the requested port / path / handle.

In the test:
 - create and start a server listening on port `0` (let the OS pick an available port)
 - create another server, set the port number explicitly to the same value as reported by the server started in the previous step
 - verify that `start()` returns a rejected promise and the error has `code` set to `EADDRINUSE` - see http://shouldjs.github.io/#assertion-rejectedwith

Once we have that test in place, you can try to use the promisifed `listen` method and see how easy/difficult it is to correctly handle errors too (compared to what we have now).

File: packages/http-server/src/http-server.ts (hunk 28 above)
Comment (bajtos) [reply packages/http-server/src/http-server.ts]: I think we can use `readonly` modifier for all properties excluding `port`, because `port` changes as a result of `start()` call. OTOH, maybe we should keep a readonly `_port` property to hold the requested port (e.g. `0`) and then write a `port` getter that will either return the actual port reported by `this._httpServer.address().port` or fall back to `this._port`.

This has an important benefit that multiple start/stop cycles of a server configured to listen on port 0 will always pick a new port. At the moment, I believe subsequent `start()` calls will pick the temporary port number assigned by the operating system for the first `start()` call.

@hacksparrow Please add a unit-test to verify the behavior of our server in this edge case.

File: packages/http-server/src/http-server.ts (hunk 31 above)
Comment (bajtos) [reply packages/http-server/src/http-server.ts]: +1 to handle IPv6 localhost URLs correctly, although I am fine with leaving that out of scope of this pull request. Isn't there any Node.js core API or perhaps an npm module that can solve this problem for us?

The built-in `url` API seems to work well

```
$ node
> require('url').format({protocol: 'http', hostname: '::1', port: 80, pathname: '/foo'})
'http://[::1]:80/foo'
```

The docs says this code is considered as legacy and should not be used in new code :(

Unfortunately the newer WhatWG API does not encode IPv6 local addressed as the old one, so I am not sure how much benefit we can get from using it.

See https://github.com/nodejs/help/issues/1176#issuecomment-375131204

```ts
const u = new (require('url').URL)('http://./');
u.protocol = 'https';
u.host = '[::1]';
u.port = 3000;
u.pathname = '/foo';
u.href 
// returns 
// 'https://[::1]:3000/foo'
```

---

-1 for Travis-specific hacks, see https://github.com/strongloop/loopback-next/pull/1369#discussion_r192014280

> We cannot ship workarounds for Travis CI in code that will be used by LB apps in production!
>
> If IPv6 localhost addresses are causing problems in Travis, then let's fix our tests to explicitly listen on 127.0.0.1 instead.

File: packages/http-server/src/http-server.ts (hunk 28 above)
Comment (hacksparrow) [reply packages/http-server/src/http-server.ts]: @raymondfeng even an undefined `host` will behave like `port` set to `0`. If the `host` value is not set during instantiation, the `HttpServer` instance will store it as `undefined`, however, the underlying `http`/`https` will generate a host value for the server - something like `::`.

Since `url` is dependent on `host` and `port`, the only property that can be implemented as a `readonly` would be `protocol`.

@bajtos I will add tests for both `host` and `port`.

File: packages/http-server/src/http-server.ts (hunk 31 above)
Comment (hacksparrow) [reply packages/http-server/src/http-server.ts]: @raymondfeng can we do the "robustification" after landing the basic functionality?

File: packages/http-server/src/http-server.ts (hunk 30 above)
Comment (hacksparrow) [reply packages/http-server/src/http-server.ts]: @bajtos added the test.

@raymondfeng why have callbacks, when we can just `await`?

File: packages/http-server/src/http-server.ts (hunk 29 above)
Comment (hacksparrow) [reply packages/http-server/src/http-server.ts]: Added.

File: packages/http-server/src/http-server.ts (hunk 28 above)
Comment (hacksparrow) [reply packages/http-server/src/http-server.ts]: Added.

File: packages/http-server/src/http-server.ts (hunk 31 above)
Comment (raymondfeng) [reply packages/http-server/src/http-server.ts]: If we force tests to only listen `127.0.0.1`, we then lose the coverage for `host === undefined`.

Comment (bajtos) [reply packages/http-server/src/http-server.ts]: Makes sense. I am proposing to write two tests for `host === undefined` case:

 - One test to verify what is returned by `host` and `url` properties. I think we don't need to start the server for this and thus these tests can be at unit-test level.
 - One test to verify that server listening on all hosts accepts connections from `127.0.1`. The trick is to build the request URL manually (`http://127.0.0.1:${server.port}/`) in the test, don't use `server.url` property.

**Comment by hacksparrow** [comment]
@slnode ok to test
